src/data-model-providers/codegen/ServerClusterInterfaceRegistry.h

```diff
@@ -0,0 +1,132 @@
+/*
+ *    Copyright (c) 2025 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <app/ConcreteClusterPath.h>
+#include <app/server-cluster/ServerClusterInterface.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/DataModelTypes.h>
+
+namespace chip {
+namespace app {
+
+/// Represents an entry in the server cluster interface registry for
+/// a specific interface.
+///
+/// In practice this is a single-linked list element.
+struct ServerClusterRegistration
+{
+    // A single-linked list of clusters registered for the given `endpointId`
+    ServerClusterInterface * serverClusterInterface = nullptr;
+    ServerClusterRegistration * next                = nullptr;
```

Comment:
Followup: we should make this protected and the registry a friend; that fact that these are linked lists is an implementation detail, not part of the API.

Reply:
I am torn on this ... generally I would rather really ban `friend` as whenever that was used, it seemed to more hinder development and testing rather than "make things safer".